codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules2/RuleSetExpression.java

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.codegen.poet.rules2;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Represents a rule set expression.
+ */
+public final class RuleSetExpression implements RuleExpression {
+    private final RuleType type;
+    private final List<RuleExpression> conditions;
+    private final List<RuleSetExpression> children;
+    private final ErrorExpression error;
+    private final EndpointExpression endpoint;
+    private final String ruleId;
+
+    RuleSetExpression(Builder builder) {
+        this.type = RuleRuntimeTypeMirror.VOID;
+        if (builder.conditions.isEmpty()) {
+            this.conditions = Collections.emptyList();
+        } else {
+            this.conditions = Collections.unmodifiableList(new ArrayList<>(builder.conditions));
+        }
+        if (!builder.children.isEmpty()) {
+            Validate.isNull(builder.error, "RuleSet with children cannot have error result");
+            Validate.isNull(builder.endpoint, "RuleSet with children cannot have endpoint result");
+            this.children = Collections.unmodifiableList(new ArrayList<>(builder.children));
+            this.error = null;
+            this.endpoint = null;
+        } else if (builder.endpoint != null) {
+            Validate.isNull(builder.error, "RuleSet with endpoint cannot have error result");
+            this.children = null;
+            this.error = null;
+            this.endpoint = builder.endpoint;
+        } else if (builder.error != null) {
+            Validate.isNull(builder.endpoint, "RuleSet with error cannot have endpoint result");
+            this.children = null;
+            this.error = builder.error;
+            this.endpoint = null;
+        } else {
+            throw new IllegalArgumentException("At exactly one of error, endpoint or tree, none given");
+        }
+        this.ruleId = builder.ruleId;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public RuleExpressionKind kind() {
+        return RuleExpressionKind.RULE_SET;
+    }
+
+    @Override
+    public StringBuilder appendTo(StringBuilder buf) {
```

Comment:
Is this correct? `toString()` would just be an empty string. There's a couple other `Expression` impls that don't append to `buf` also